pkg/handlers/v1/nexposeassetproducer.go

```diff
@@ -47,8 +47,10 @@ func (h *NexposeScannedAssetProducer) Handle(ctx context.Context, in ScanInfo) {
 					defer wg.Done()
 					err := h.Producer.Produce(ctx, asset)
 					if err != nil {
+						reason := err.Error()
+						stater.Count("producerfailure", 1, fmt.Sprintf("site:%s", in.SiteID), reason)
```

Comment:
I'm pretty questionable about passing reason as a tag to be emitted, feels like a bad idea... Thoughts?

Reply:
You're right to question it. We should never pass information like this through metrics. Detailed debug and error data should be passed through logs where a human can review or search through the information later. The purpose of a metrics tag is to enable the filtering and grouping of data. This means all tags values must be finite, enumerable, and have a relatively low number of possible values. The number of possible tags values is often discussed in terms of cardinality because they form a set. High cardinality sets are a hard problem to work with and most metrics systems don't handle them well.

Personally, I think having the siteID is already pushing it because we can't guarantee that there are a bounded number of values in all usage of Nexpose. We definitely cannot pass arbitrary string content in, though.

Reply:
But this ticket calls for emitting stats where we fail to produce on assets, tagged with a siteID and a reason of failure. As you said, I think siteID should be ok, but how would I emit a reason for failure if it's only in terms of a string? Like you said, it seems like a bad idea, but then should we reevaluate the ticket?

Reply:
If you want to add a reason tag, you could select it from a finite list of tokens defined as constants, e.g. 
```
const (
  errReasonBlacklist = "blacklist"
  errReasonFieldMissing = "field-missing"
  errReasonProducerFailure = "producer-failure"
  errReasonUnknown = "unknown"
)
...
err := h.Producer.Produce(ctx, asset)
if err != nil {
  stater.Count("producerfailure", 1, fmt.Sprintf("site:%s", in.SiteID), fmt.Sprintf("reason:%s", errReasonProducerFailure))
```

Reply:
One option is to conform the string into the rules I mentioned: `finite, enumerable, and have a relatively low number of possible values`. For example, you can use a type switch to choose a token:
```golang
reason := ""
switch err.(type) {
  case domain.SomeError:
    reason = "some_error"
  case domain.SomeOtherError:
    reason = "other_error"
  default:
    reason = "UNKNOWN"
}
```

However, you don't really have any way of predicting the possible values of the error because you don't control the errors that the producer returns. Virtually all the errors would have the reason of "UNKNOWN". Additionally, the error types returned will differ depending on which driver is selected for the producer. For example, you'll see HTTP related errors then using the POST version, AWS error types when using the BENTHOS version, and no error types when using the NULL verison.

Personally, I don't see the value in adding a reason to the metric if there aren't a specific, known set of values we want to track over time. The value of a tag should never be a surprise and we should not use tags to discover what the possible values of a variable are. A better tool for that would be the logs. Logs are where we should capture detailed exception data for analysis by a human later.